### ISSUE TITLE
Allow defining variables into PromQL, so I can make my queries shorter and understandable

### ISSUE THREAD

**Issue by aSemy** [prometheus/prometheus#11609]
### Proposal

Currently writing a PromQL query longer than a couple of lines quickly becomes extremely frustrating. This is because filters and queries can't be re-used, and have to be copy and pasted, which quickly becomes very messy. Queries can also become highly nested, and it's hard to see where brackets start and end.

For example, this query measures the number of seconds the Apdex score is below 0.8 (I think)

```promql
sum_over_time((sum without (client) (
  label_replace(
    (
          (
              (
                  sum by (client, uri) (rate(feign_Client_seconds_bucket{env="prod",le="1"}[5m]))
                +
                  sum by (client, uri) (rate(feign_Client_seconds_bucket{env="prod",le="4"}[5m]))
              ) > 0
          )
        / sum by (client, uri) (rate(feign_Client_seconds_count{env="prod"}[5m]))
      - 1
    ),
    "client_name",
    "$1",
    "client",
    ".*?([^\\.]*)"
  )
) < bool 0.8)[1h:1s])
```

I would like to be able to re-write this so it's more legible, re-usable, and compartmentalised.

Here's a very example psuedocode (based on Kotlin syntax) that I've commented to explain the purpose.

```kotlin
// define a constant interval, which can be used in all queries
val baseInterval = 5m

// create a custom filter, that can be re-used. This makes updating the filter much more easy.
val clientFilter = {env="prod", client="my.app.service.SomeClient"}

// store metrics in variables, re-using the custom filter. The variable names help document the query.
val rateClientDurationUnder1Second = sum by (client, uri) (
                             // clientFilter is re-used, but additional filters can be applied
  rate(feign_Client_seconds_bucket{filter=clientFilter, le="1"}[baseInterval])
)

val rateClientDurationUnder4Seconds = sum by (client, uri) (
  rate(feign_Client_seconds_bucket{filter=clientFilter, le="4"}[baseInterval])
)

val clientDurationTotalCount = sum by (client, uri) (
  rate(feign_Client_seconds_count{filter=clientFilter}[baseInterval])
)

// create custom functions that can be re-used
fun renameClientLabel(metric: Metric) = 
  sum without (client) (
    label_replace(metric, "client_name", "$1", "client", ".*?([^\\.]*)")
  )

/** Count the number of seconds per hour the provided metric is below 0.8 */
fun sumSecondsMetricBelowThresholdPerHour(metric: Metric) = 
  sum_over_time(metric < bool 0.8)[1h:1s])

// now the actual query becomes a little more succinct

sumSecondsMetricBelowThresholdPerHour(
    (
       ((rateClientDurationUnder1Second + rateClientDurationUnder4Seconds) > 0)
        / clientDurationTotalCount
    ) - 1 // subtract 1, because two buckets were summed)
)
```

The clarity afforded by variable names, and being able to control the query by changing the values in a single place of code, makes the second option much more clear and easy to use.

**Comment by beorn7** [comment]
This is somewhat related to the recent dev-summit discussion about specifying common labels (see [the dev-summit notes](https://docs.google.com/document/d/11LC3wJcVk00l8w5P3oLQ-m3Y37iom6INAMEu2ZAGIIE/edit?pli=1) and search for "RichiH: Common labels for queries", sorry for not being able to provide a deep link).

Note that the decision to not allow user-defined functions or macro expansions as part of PromQL was a very deliberate one (based on years of experience with other query languages). While the addition of some "common label" syntax is consensus according to the dev-summit, I would not expect the same for advanced features like the one you have sketched out above. It is a good practice, however, to use an external configuration system and/or macro processor to generate PromQL, theredoy adding those features in a layer on top of PromQL itself. (Cf. the mixins written in Jsonnet, or many other solutions based on Helm, M4, …)

**Comment by aSemy** [comment]
Common labels would certainly help. 

For reference, and to try and protect against link-rot, here's relevant part copy and pasted

> #### RichiH: Common labels for queries
> * VictoriaQL supports [`WITH ( commonFilters = {instance=~"$node:$port",job=~"$job"} )`](https://play.victoriametrics.com/promql/expand-with-exprs), allowing users to write relevant labels only once
> * Seems useful; even if nesting could be a foot-gun
> * Custom functions seem like a larger debate, even though subqueries exist and might get us most of the way there
> * RichiH: I don’t  like VQL’s syntax, but VQL sets a precedent. If we end up doing anything in this direction, we should discuss and ideally coordinate with VM.
>   * What about a `common_labels()` function?
> * Beorn: Note that the ViMe WITH feature is extremely powerful. While that sounds great at first, it can be a real footgun for the monitoring use case. I understand that this item is just about the “common labels” functionality and not about the more powerful aspects of ViMe’s WITH. Just saying this here to make sure we all know what we are talking about.
> Richi: only allow labels, WITH is too powerful and too overloaded, we should be deliberate about every single extension we allow
> * Chris M: What if we you have more labels in other places?
> * RichiH: Build a tree of PromQL execution, every time you add more labels, you just add them. At the end, you concat all labels
> * Goutham: Make it part of the API?
> * Richih: No, needs to be part of PromQL. I don’t want this to work in Grafana, and not in Prometheus. Also, I used to run a lot of queries against Prometheus directly and work on the JSON
> * Tom: Feel like better join syntax with label-rewriting/matching on different labels names is higher priority.  But we have consensus to consider a proposal here already, someone just needs to do it.
> 
> CONSENSUS: We want to support common labels as part of PromQL

**Comment by alexgreenbank** [comment]
Hello from the bug scrub.

Fixed title and added appropriate labels.

**Comment by ringerc** [comment]
I threw around some syntax suggestions in https://github.com/prometheus/prometheus/issues/15368 before realising this ticket already covered pretty much the same thing.

**Comment by bobrik** [comment]
This query:

```
container_memory_usage_bytes{app_name="prometheus.service"} + container_memory_swap
```

Is 30x slower than this equivalent query:

```
container_memory_usage_bytes{app_name="prometheus.service"} + container_memory_swap{app_name="prometheus.service"}
```

It is related to this issue, but I would argue that both should be fast _without_ any extra syntax, just by applying the intersection of label selectors. Should I make a separate issue for that?

**Comment by beorn7** [comment]
There is already a separate issue for that: https://github.com/prometheus/prometheus/issues/8053
